plugins/src/test/java/opengrok/auth/plugin/LdapServerTest.java

```diff
@@ -0,0 +1,96 @@
+package opengrok.auth.plugin;
+
+import opengrok.auth.plugin.ldap.LdapServer;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.util.concurrent.CountDownLatch;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+
+public class LdapServerTest {
+
+    @Test
+    public void testInvalidURI() {
+        LdapServer server = new LdapServer("foo:/\\/\\foo.bar");
+        assertFalse(server.isReachable());
+    }
+
+    @Test
+    public void testGetPort() throws URISyntaxException {
+        LdapServer server = new LdapServer("ldaps://foo.bar");
+        assertEquals(636, server.getPort());
+
+        server = new LdapServer("ldap://foo.bar");
+        assertEquals(389, server.getPort());
+
+        server = new LdapServer("crumble://foo.bar");
+        assertEquals(-1, server.getPort());
+    }
+
+    @Test
+    public void testSetGetUsername() {
+        LdapServer server = new LdapServer();
+
+        assertNull(server.getUsername());
+        assertNull(server.getPassword());
+
+        final String testUsername = "foo";
+        server.setUsername(testUsername);
+        assertEquals(testUsername, server.getUsername());
+
+        final String testPassword = "bar";
+        server.setPassword(testPassword);
+        assertEquals(testPassword, server.getPassword());
+    }
+
+    @Test
+    public void testIsReachable() throws UnknownHostException, InterruptedException, URISyntaxException {
+        // Start simple TCP server on port 6336. It has to be > 1024 to avoid BindException
+        // due to permission denied.
+        int testPort = 6336;
+        InetAddress localhostAddr = InetAddress.getLocalHost();
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        Thread thread = new Thread(() -> {
+            try {
+                ServerSocket socket = new ServerSocket(testPort, 1, localhostAddr);
+                startLatch.countDown();
+                Socket client = socket.accept();
+                client.close();
+                socket.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        });
+
+        // TODO:
+        //  There is still a tiny window between when the latch unblocks and the server actually starts accept().
```

Comment:
Any way to make reliable or at least to await a finite time. E.g. what if `new ServerSocket` throws, and latch is never decremented.

Reply:
sure, good idea.